k2/csrc/cuda/context.h

```diff
@@ -28,7 +28,7 @@ constexpr DeviceType kCpu = DeviceType::kCpu;
 
 class Context;
 using ContextPtr = std::shared_ptr<Context>;
-constexpr cudaStream_t kCudaStreamInvalid = (cudaStream_t)(~((size_t)0));
+const cudaStream_t kCudaStreamInvalid = (cudaStream_t)(~((size_t)0));
```

Comment:
I change this `constexpr` to `const` because some compilers will complain: it's an error to `use reinterpret_cast` in a constexpr.

At least, the compiler (GCC 7.5.0) from the colab notebook will complain.

Reply:
OK, seems the latest GCC version has strong requirement. On my local machine (Gcc 4.8.5), when I build the whole project, there are only warnings from `debug.h`.

Reply:
I just thought maybe we can add static?

Reply:
Yes, I think we can.

`constexpr` implies `const` which implies internal linkage, e.g., `static`.

---

For the following test program `test.cc`
```cpp
const int a = 10;
int test() { return a + 2; }
```

If you use
```bash
g++ -c test.cc
nm test.o
```
you can see that the symbol `a` has an internal linkage

```
0000000000000000 T _Z4testv
0000000000000000 r _ZL1a
```

If you use `g++ -O2 -c test.cc`, the symbol `a` is optimized out.

Reply:
Perhaps it can be made an enum or even a #define expression?  Problem with a variable is the link problems get very complicated, I dont know how to define variables in headers and not get link problems.

Reply:
Let's do it as `#define` like NVidia does, e.g. `#define cudaStreamPerThread ((cudaStream_t)0x2). `
It's a shame the C++ standard doesn't let us do this very simple thing...

Reply:
> Let's do it as #define like NVidia does

Done.

Reply:
> I dont know how to define variables in headers and not get link problems

variables in a header file with an internal linkage will not cause link problems, that is, you can define
static or const variables in a header file, which are commonly used for primitive types.

Reply:
> It's a shame the C++ standard doesn't let us do this very simple thing...

`cudaStream_t` is a **pointer** to a struct. The statement involves `reinterpret_cast`, which is not that simple.

According to <https://en.cppreference.com/w/cpp/language/constant_expression>:

> Core constant expressions
> A core constant expression is any expression whose evaluation would not evaluate any one of the following:
>  ....
> 16. (until C++20)dynamic_cast
> 17. reinterpret_cast